### PR TITLE
Enhancement for EHPA recommendation

### PR DESCRIPTION
**Pull request by qmhu**
#### What type of PR is this?
feature && enhancement

#### What this PR does / why we need it:

enhancement for ehpa recommendation
implement recommendation updater

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: pkg/recommend/advisor/ehpa.go [hunk 1]
```diff
@@ -141,25 +152,186 @@ func (a *EHPAAdvisor) Name() string {
...
+			cpuUsagePredictionMap[sampleTime.Hour()] = append(cpuUsagePredictionMap[sampleTime.Hour()], sample.Value)
+		} else {
+			newUsageInHour := make([]float64, 0)
+			newUsageInHour = append(newUsageInHour, sample.Value)
+			cpuUsagePredictionMap[sampleTime.Hour()] = newUsageInHour
+		}
+	}
+
+	// use mean to deburring data
+	var meanUsages []float64
```
Comment (yufeiyu) [inline pkg/recommend/advisor/ehpa.go:190]: 'mean' is confusing, cause actually what is used here is median.

File: pkg/recommend/advisor/ehpa.go [hunk 2]
```diff
@@ -141,25 +152,186 @@ func (a *EHPAAdvisor) Name() string {
...
+	for _, usageInHour := range cpuUsagePredictionMap {
+		meanUsage, err := stats.Median(usageInHour)
+		if err != nil {
+			return err
+		}
+		meanUsages = append(meanUsages, meanUsage)
+	}
+
+	meanMax := math.SmallestNonzeroFloat64
```
Comment (yufeiyu) [inline pkg/recommend/advisor/ehpa.go:199]: medianMax/medianMin is better?

File: pkg/recommend/advisor/ehpa.go [hunk 3]
```diff
@@ -141,25 +152,186 @@ func (a *EHPAAdvisor) Name() string {
...
+	return minReplicas, nil
+}
+
+// proposeMaxReplicas use max cpu usage to compare with target pod cpu usage to get the max replicas.
+func (a *EHPAAdvisor) proposeMaxReplicas(cpuUsages []float64, targetUtilization int32, minReplicas int32) (int32, error) {
+	maxReplicasFactor, err := strconv.ParseFloat(a.Context.ConfigProperties["ehpa.max-replicas-factor"], 64)
+	if err != nil {
+		return 0, err
+	}
+	// use percentile to deburring data
```
Comment (yufeiyu) [inline pkg/recommend/advisor/ehpa.go:302]: typo?

File: pkg/recommend/advisor/ehpa.go [hunk 4]
```diff
@@ -141,25 +152,186 @@ func (a *EHPAAdvisor) Name() string {
...
+	if err != nil {
+		return 0, err
+	}
+	requestsPod, err := utils.CalculatePodTemplateRequests(a.PodTemplate, corev1.ResourceCPU)
+	if err != nil {
+		return 0, err
+	}
+
+	// request * targetUtilization is the target average cpu usage, use total p95thCpu to divide, we can get the expect max replicas.
+	calcMaxReplicas := (p95thCpu * 100 * 1000 * maxReplicasFactor) / float64(int32(requestsPod)*targetUtilization)
```
Comment (yufeiyu) [inline pkg/recommend/advisor/ehpa.go:313]: why *100?

File: pkg/recommend/advisor/ehpa.go (hunk 3 above)
Comment (qmhu) [reply pkg/recommend/advisor/ehpa.go:302]: deburring -> 去毛刺

File: pkg/recommend/advisor/ehpa.go (hunk 4 above)
Comment (qmhu) [reply pkg/recommend/advisor/ehpa.go:313]: targetUtilization is [1-100]

**Pull request merged**